### PR TITLE
Scala3 support patch 1

### PR DESCRIPTION
**Pull request by andreaTP**
The first batch of unharmful patches toward Scala 3 compatibility.
If the CI is happy everything should be good 🙂 

From the findings and discussions in #29929 

Special thanks to @smarter for the support!

### CODE REVIEW

File: project/Dependencies.scala [hunk 1]
```diff
@@ -32,7 +32,7 @@ object Dependencies {
 
   val sslConfigVersion = "0.4.2"
 
-  val scalaTestVersion = "3.1.4"
+  val scalaTestVersion = "3.2.3"
```
Comment (raboof) [inline project/Dependencies.scala:35]: looks like org.scalatestplus:mockito-3-3 was not released for scalaTest 3.2.3 (https://repo1.maven.org/maven2/org/scalatestplus/mockito-3-3_2.12/) - perhaps we should update to org.scalatestplus:mockito-3-4?

Comment (raboof) [reply project/Dependencies.scala:35]: there's been some discussion on when to update to scalatest 3.2 in #29368

Comment (andreaTP) [reply project/Dependencies.scala:35]: Thanks for the context @raboof !
Two options for moving forward:
 - I revert back the bump of ScalaTest for having this PR merged
 - we are in 2021 and we can possibly unblock the bump since @patriknw said:
  > we should be careful to not upset users with forcing another update this soon again

as you prefer.

**Comment by raboof** [comment]
OK TO TEST

File: akka-actor-tests/src/test/scala/akka/actor/TypedActorSpec.scala [hunk 2]
```diff
@@ -72,7 +72,7 @@ object TypedActorSpec {
 
     @silent
     @throws(classOf[TimeoutException])
-    def self = TypedActor.self[Foo]
+    def self = akka.actor.TypedActor.self[Foo]
```
Comment (patriknw) [inline akka-actor-tests/src/test/scala/akka/actor/TypedActorSpec.scala:75]: not that I care about this test, but why is this needed? isn't it in the same package?

File: akka-actor/src/main/scala/akka/actor/ActorPath.scala [hunk 3]
```diff
@@ -87,7 +87,7 @@ object ActorPath {
   final def validatePathElement(element: String, fullPath: String): Unit = {
     def fullPathMsg = if (fullPath ne null) s""" (in path [$fullPath])""" else ""
 
-    (findInvalidPathElementCharPosition(element): @switch) match {
+    (findInvalidPathElementCharPosition(element)) match {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/ActorPath.scala:90]: Does this make the match less efficient for Scala 2.12 and 2.13?
I'm not sure if this is used in any hot path  but it could be since it's validating actor names. We use switch  at more places so this might become a bigger problem if switch isn't supported.

File: project/Dependencies.scala (hunk 1 above)
Comment (patriknw) [reply project/Dependencies.scala:35]: can we use a different version of ScalaTest for Scala 3?

Comment (andreaTP) [reply project/Dependencies.scala:35]: no, `3.2.3` is the only single release of scalatest published for Scala 3.
https://repo1.maven.org/maven2/org/scalatest/scalatest-core_3.0.0-M3/

File: akka-actor-tests/src/test/scala/akka/actor/TypedActorSpec.scala (hunk 2 above)
Comment (andreaTP) [reply akka-actor-tests/src/test/scala/akka/actor/TypedActorSpec.scala:75]: In this file, Scala 3 is complaining about the ambiguous reference to `TypedActor`.
In between this `object` and a local `class` that comes from somewhere.

File: akka-actor/src/main/scala/akka/actor/ActorPath.scala (hunk 3 above)
Comment (andreaTP) [reply akka-actor/src/main/scala/akka/actor/ActorPath.scala:92]: Please note that I haven't removed all the `@switch` cases in the codebase but only 2.
In those cases the `switch` doesn't have enough cases and is not going to be optimized, Scala 2 silently accept the code while Scala 3 emits an error.

I added a comment with a note for the future, hope is enough.

File: project/Dependencies.scala (hunk 1 above)
Comment (patriknw) [reply project/Dependencies.scala:35]: Sorry, my question was unclear. I mean, can we use ScalaTest 3.1 for Scala 2 and ScalaTest 3.2 for Scala 3?

Comment (andreaTP) [reply project/Dependencies.scala:35]: well, yes, this is an option.
If this unblocks this PR I can do that, but I think is much cleaner for Akka to keep using a single version of ScalaTest.
Also, I will probably have to switch also the Mockito and ScalaCheck versions, is it acceptable?

Comment (patriknw) [reply project/Dependencies.scala:35]: I don't think we have the bandwidth to investigate what an incompatible bumb of ScalaTest would mean for our Scala 2.12 and 2.13 users and especially the full Akka ecosystem of dependent libraries at this moment.
Quickest way forward for this PR would be to use a 3.2.3 only for Scala 3.

Comment (andreaTP) [reply project/Dependencies.scala]: Ok, reverted the bump of ScalaTest and ScalaCheck (will do the switch after #29932 to avoid conflicts).
Kept the bump of Mockito though.

**Comment by patriknw** [comment]
Superseded by https://github.com/akka/akka/pull/29956, I assume
